File: app/src/main/java/com/xjm/xxd/dribbird/account/TokenManager.java
package com.xjm.xxd.dribbird.account;

import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;
import com.xjm.xxd.dribbird.api.ApiConstants;
import com.xjm.xxd.dribbird.api.okhttp.OkHttpManager;
import com.xjm.xxd.dribbird.utils.SharedPreferConstants;
import com.xjm.xxd.dribbird.utils.SharedPreferencesUtils;

import java.util.HashMap;
import java.util.Map;

import okhttp3.Response;
import okhttp3.ResponseBody;

import static com.xjm.xxd.dribbird.api.ApiConstants.EQUAL_MARK;

/**
 * @author : xiaoxiaoda
 *         date: 16-11-29
 *         email: daque@hustunique.com
 */
public class TokenManager {

    private static String mOAuthUrl;

    /**
     * build oauth2 authentication url
     *
     * @return the url to auth
     */
    public static String getOAuth2Url() {
        if (TextUtils.isEmpty(mOAuthUrl)) {
            StringBuilder builder = new StringBuilder(ApiConstants.OAUTH_BASE_URL);
            builder.append(ApiConstants.AUTHORIZE)
                    .append(ApiConstants.QUESTION_MARK)
                    .append(ApiConstants.CLIENT_ID)
                    .append(EQUAL_MARK)
                    .append(ApiConstants.DRIBBLE_CLIENT_ID)
                    .append(ApiConstants.AND_MARK)
                    .append(ApiConstants.REDIRECT_URI)
                    .append(EQUAL_MARK)
                    .append(ApiConstants.DEFAULT_REDIRECT_URI_PRE);
            mOAuthUrl = builder.toString();
        }
        return mOAuthUrl;
    }

    public static boolean isOAuth2Url(String targetStr) {
        return !TextUtils.isEmpty(targetStr) && getOAuth2Url().equals(targetStr);
    }

    public static boolean isMatchRedirectUrl(String targetStr) {
        return !TextUtils.isEmpty(targetStr) && targetStr.startsWith(ApiConstants.DEFAULT_REDIRECT_URI_PRE);
    }

    /**
     * return if there is a local token
     * @return
     */
    public static boolean hasAvaliableToken() {
        String tokenBean = SharedPreferencesUtils.getDefaultString(SharedPreferConstants.TOKEN_BEAN, "");
        if (TextUtils.isEmpty(tokenBean)) {
            return false;
        }
        Gson gson = OkHttpManager.getInstance().getGson();
        try {
            gson.fromJson(tokenBean, TokenBean.class);
        } catch (JsonSyntaxException e) {
            e.printStackTrace();
            return false;
        }
        return true;
    }

    public static
    @Nullable
    TokenBean requestForToken(String returnCode) {
        if (TextUtils.isEmpty(returnCode)) {
            return null;
        }
        Map<String, String> params = new HashMap<>();
        params.put(ApiConstants.CLIENT_ID, ApiConstants.DRIBBLE_CLIENT_ID);
        params.put(ApiConstants.CLIENT_SECRET, ApiConstants.DRIBBLE_CLIENT_SECRET);
        params.put(ApiConstants.CODE, returnCode);
        Response response = OkHttpManager.getInstance().postSync(
                ApiConstants.OAUTH_BASE_URL + ApiConstants.TOKEN,
                params);
        if (response == null) {
            return null;
        }
        ResponseBody responseBody = response.body();
        if (responseBody == null) {
            return null;
        }
        Gson gson = OkHttpManager.getInstance().getGson();
        TokenBean tokenBean = gson.fromJson(responseBody.charStream(), TokenBean.class);
        SharedPreferencesUtils.setDefaultString(SharedPreferConstants.TOKEN_BEAN, gson.toJson(tokenBean));
        return tokenBean;
    }

}

File: app/src/main/java/com/xjm/xxd/dribbird/account/TokenManager.kt
package com.xjm.xxd.dribbird.account

import android.net.Uri
import com.google.gson.JsonSyntaxException
import com.xjm.xxd.dribbird.network.*
import com.xjm.xxd.framework.network.GsonManager
import com.xjm.xxd.framework.network.OkHttpManager
import com.xjm.xxd.framework.data.CommonStore
import com.xjm.xxd.skeleton.kotlinext.nullAsFalse

/**
 * @author : dada
 * date: 16-11-29
 * email: daque@hustunique.com
 */

object TokenManager {

    private var mOAuthUrl: String = ""

    private var mTokenBean: TokenBean? = null

    /**
     * build oauth2 authentication url
     * @return the url to auth
     */
    val oAuth2Url: String
        get() {
            if (mOAuthUrl.isEmpty()) {
                val builder = StringBuilder(OAUTH_BASE_URL)
                builder.append(AUTHORIZE)
                        .append(QUESTION_MARK)
                        .append(CLIENT_ID)
                        .append(EQUAL_MARK)
                        .append(DRIBBLE_CLIENT_ID)
                        .append(AND_MARK)
                        .append(REDIRECT_URI)
                        .append(EQUAL_MARK)
                        .append(DEFAULT_REDIRECT_URI_PRE)
                mOAuthUrl = builder.toString()
            }
            return mOAuthUrl
        }

    val tokenBean: TokenBean?
        get() {
            if (mTokenBean == null) {
                val tokenBean = CommonStore.currentTokenBean
                if (tokenBean.isEmpty()) {
                    return null
                }
                val gson = OkHttpManager.instance.gson
                try {
                    mTokenBean = gson.fromJson(tokenBean, TokenBean::class.java)
                } catch (e: JsonSyntaxException) {
                    e.printStackTrace()
                    return null
                }
            }
            return mTokenBean
        }

    fun isOAuth2Url(targetStr: String?): Boolean {
        return !targetStr.isNullOrEmpty() && targetStr.equals(oAuth2Url)
    }

    fun isMatchRedirectUrl(targetUrl: String?): Boolean {
        return !targetUrl.isNullOrEmpty() && targetUrl?.startsWith(DEFAULT_REDIRECT_URI_PRE).nullAsFalse()
    }

    fun isMatchRedirectUrl(targetUri: Uri?): Boolean {
        return targetUri != null && targetUri.toString().startsWith(DEFAULT_REDIRECT_URI_PRE)
    }

    /**
     * return if there is a local token
     * @return
     */
    fun hasAvailableToken(): Boolean {
        val tokenBean = CommonStore.currentTokenBean
        if (tokenBean.isEmpty()) {
            return false
        }
        val gson = OkHttpManager.instance.gson
        try {
            mTokenBean = gson.fromJson(tokenBean, TokenBean::class.java)
        } catch (e: Throwable) {
            e.printStackTrace()
            return false
        }

        return true
    }

    fun requestForToken(returnCode: String?): TokenBean? {
        if (returnCode.isNullOrEmpty()) {
            return null
        }
        val params = hashMapOf<String, String>()
        params[CLIENT_ID] = DRIBBLE_CLIENT_ID
        params[CLIENT_SECRET] = DRIBBLE_CLIENT_SECRET
        params[CODE] = returnCode!!
        val response = OkHttpManager.instance.postSync(
                OAUTH_BASE_URL + TOKEN,
                params) ?: return null
        val responseBody = response.body() ?: return null
        return GsonManager.gson().fromJson(responseBody.charStream(), TokenBean::class.java)
    }

}
